listenbrainz/webserver/static/js/src/APIService.ts

```diff
@@ -0,0 +1,191 @@
+    const userNamesForQuery: string = userNames.join(",");
+
+    let query = `${this.APIBaseURI}/users/${userNamesForQuery}/recent-listens`;
+
+    if (limit) {
+      query += `?limit=${limit}`;
+    }
+
+    const response = await fetch(query, {
+      method: "GET",
+    });
+    this.checkStatus(response);
+    const result = await response.json();
+
+    return result.payload.listens;
+  };
+
+  getListensForUser = async (
+    userName: string,
+    minTs?: number,
+    maxTs?: number,
+    count?: number
+  ): Promise<Array<Listen>> => {
+    if (maxTs && minTs) {
+      throw new SyntaxError(
+        "Cannot have both minTs and maxTs defined at the same time"
+      );
+    }
+
+    let query: string = `${this.APIBaseURI}/user/${userName}/listens`;
+
+    const queryParams: Array<string> = [];
+    if (maxTs) {
+      queryParams.push(`max_ts=${maxTs}`);
+    }
+    if (minTs) {
+      queryParams.push(`min_ts=${minTs}`);
+    }
+    if (count) {
+      queryParams.push(`count=${count}`);
+    }
+    if (queryParams.length) {
+      query += `?${queryParams.join("&")}`;
+    }
+
+    const response = await fetch(query, {
+      method: "GET",
+    });
+    this.checkStatus(response);
+    const result = await response.json();
+
+    return result.payload.listens;
+  };
+
+  refreshSpotifyToken = async (): Promise<string> => {
+    const response = await fetch("/profile/refresh-spotify-token", {
+      method: "POST",
+    });
+    this.checkStatus(response);
+    const result = await response.json();
+    return result.user_token;
+  };
+
+  /*
+    Send a POST request to the ListenBrainz server to submit a listen
+  */
+  submitListens = async (
+    userToken: string,
+    listenType: ListenType,
+    payload: SubmitListensPayload
+  ): Promise<Response> => {
+    if (JSON.stringify(payload).length <= this.MAX_LISTEN_SIZE) {
+      // Payload is within submission limit, submit directly
+      const struct = {
+        listen_type: listenType,
+        payload,
+      };
+
+      const url = `${this.APIBaseURI}/submit-listens`;
+
+      try {
+        const response = await fetch(url, {
+          method: "POST",
+          headers: {
+            Authorization: `Token ${userToken}`,
+            "Content-Type": "application/json;charset=UTF-8",
+          },
+          body: JSON.stringify(struct),
+        });
+
+        // we skip listens if we get an error code that's not a rate limit
+        if (response.status === 429) {
+          // Rate limit error, this should never happen, but if it does, try again in 3 seconds.
+          setTimeout(
+            () => this.submitListens(userToken, listenType, payload),
+            3000
+          );
+        }
+        return response; // Return response so that caller can handle appropriately
+      } catch {
+        // Retry if there is an network error
+        setTimeout(
+          () => this.submitListens(userToken, listenType, payload),
+          3000
+        );
```

Comment:
Something should be returned otherwise extra calls would be made below

Reply:
Should be fixed as a part of this ticket: https://tickets.metabrainz.org/browse/LB-516